File: app/src/main/java/com/github/goldy1992/mp3player/client/activities/SplashScreenEntryActivity.java
package com.github.goldy1992.mp3player.client.activities;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import androidx.appcompat.app.AppCompatActivity;

import com.github.goldy1992.mp3player.R;
import com.github.goldy1992.mp3player.client.MediaBrowserAdapter;
import com.github.goldy1992.mp3player.client.MediaBrowserConnectorCallback;
import com.github.goldy1992.mp3player.client.PermissionGranted;
import com.github.goldy1992.mp3player.client.PermissionsProcessor;
import com.github.goldy1992.mp3player.dagger.components.DaggerMediaActivityCompatComponent;
import com.github.goldy1992.mp3player.dagger.components.MediaActivityCompatComponent;
import com.github.goldy1992.mp3player.dagger.components.SplashScreenEntryActivityComponent;

import org.apache.commons.lang3.exception.ExceptionUtils;

import javax.inject.Inject;

import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;
import static com.github.goldy1992.mp3player.commons.Constants.ONE_SECOND;

public class SplashScreenEntryActivity extends AppCompatActivity
    implements  MediaBrowserConnectorCallback,
                PermissionGranted {

    private MediaBrowserAdapter mediaBrowserAdapter;
    private static final String LOG_TAG = "SPLSH_SCRN_ENTRY_ACTVTY";
    private static final long WAIT_TIME = 3000L;
    private PermissionsProcessor permissionsProcessor;
    private volatile boolean splashScreenFinishedDisplaying = false;
    private volatile boolean permissionGranted = false;
    private Intent mainActivityIntent;
    public static final int APP_TERMINATED = 0x78;
    private Thread thread = new Thread(() -> init());

    @Override
    public void onCreate(Bundle savedInstanceState) {

        if (!isTaskRoot()
                && getIntent().hasCategory(Intent.CATEGORY_LAUNCHER)
                && getIntent().getAction() != null
                && getIntent().getAction().equals(Intent.ACTION_MAIN)) {

            finish();
            return;
        }

        initialiseDependencies();
        super.onCreate(savedInstanceState);
        // TODO: have this injected so that a test implementation can be provided
        mainActivityIntent = new Intent(getApplicationContext(), getMainActivityClass());
        setContentView(R.layout.splash_screen);
        thread.start();
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.i(LOG_TAG, "onStart");
    }

    private synchronized void splashScreenRun() {
    //    Log.i(LOG_TAG, "splashscreen run");
        try {
            wait(WAIT_TIME);
        } catch (InterruptedException ex) {
            Log.e(LOG_TAG, ExceptionUtils.getMessage(ex.fillInStackTrace()));
            Thread.currentThread().interrupt();
        } finally {
            setSplashScreenFinishedDisplaying(true);
            notifyAll();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        //   Log.i(LOG_TAG, "permission result");
        boolean permissionIsGranted = false;
        if (permissions.length > 0 && grantResults.length > 0) {
            for (int i = 0; i < permissions.length; i++) {
                if (WRITE_EXTERNAL_STORAGE.equals(permissions[i]) &&
                        grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    permissionIsGranted = true;
                }
            }
        }
        if (permissionIsGranted) {
            onPermissionGranted();
        } else {
            Toast.makeText(getApplicationContext(), "Permission denied, please enable Storage permissions in settings in order to uer the app", Toast.LENGTH_LONG).show();
            finish();
        }
    }

    @Override // MediaBrowserConnectorCallback
    public void onConnected() {
        onProcessingComplete();
        Log.i(LOG_TAG, "hit on connected");

    }

    @Override// MediaBrowserConnectorCallback
    public void onConnectionSuspended() {    }

    @Override // MediaBrowserConnectorCallback
    public void onConnectionFailed() {
        Log.i(LOG_TAG, "connection failed");
    }

    private synchronized void onProcessingComplete() {
        Log.i(LOG_TAG, "processing complete");
        while (!isSplashScreenFinishedDisplaying() || !isPermissionGranted()) {
            try {
                wait(ONE_SECOND);
            } catch (InterruptedException ex) {
                String error = ExceptionUtils.getMessage(ex.fillInStackTrace());
                Log.e(LOG_TAG, error);
                Thread.currentThread().interrupt();
            }
        }
        startMainActivity(mainActivityIntent);
    }

    private void startMainActivity(Intent mainActivityIntent) {
        Log.i(LOG_TAG, "start main activity");
        startActivityForResult(mainActivityIntent, APP_TERMINATED);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == APP_TERMINATED) {
            if (mediaBrowserAdapter != null) {
                mediaBrowserAdapter.disconnect();
            }
            finish();
        }
    }

    @Override
    public void onPermissionGranted() {
        Log.i(LOG_TAG, "permission granted");
        setPermissionGranted(true);
        Runnable r = new Thread(() -> mediaBrowserAdapter.connect());
        runOnUiThread(r);
    }

    public void init() {
        Log.i(LOG_TAG, "reset");
        permissionsProcessor.requestPermission(WRITE_EXTERNAL_STORAGE);
        Thread splashScreenWaitThread = new Thread(() -> splashScreenRun());
        splashScreenWaitThread.start();
    }

    @VisibleForTesting
    public Thread getThread() {
        return thread;
    }

    @VisibleForTesting
    public boolean isSplashScreenFinishedDisplaying() {
        return splashScreenFinishedDisplaying;
    }

    @VisibleForTesting
    public boolean isPermissionGranted() {
        return permissionGranted;
    }
    @VisibleForTesting
    public void setSplashScreenFinishedDisplaying(boolean splashScreenFinishedDisplaying) {
        this.splashScreenFinishedDisplaying = splashScreenFinishedDisplaying;
    }
    @VisibleForTesting
    public void setPermissionGranted(boolean permissionGranted) {
        this.permissionGranted = permissionGranted;
    }

    @Inject
    public void setPermissionsProcessor(PermissionsProcessor permissionsProcessor) {
        this.permissionsProcessor = permissionsProcessor;
    }

    @Inject
    public void setMediaBrowserAdapter(MediaBrowserAdapter mediaBrowserAdapter) {
        this.mediaBrowserAdapter = mediaBrowserAdapter;
    }

    private void initialiseDependencies() {
        MediaActivityCompatComponent component = DaggerMediaActivityCompatComponent
            .factory()
            .create(getApplicationContext(),"SPSH_SCRN_ACTVTY_WRKR", this);
        SplashScreenEntryActivityComponent splashScreenEntryActivityComponent =
                component.splashScreenEntryActivity().create(this, this);
        splashScreenEntryActivityComponent.inject(this);
    }

    Class<?> getMainActivityClass() {
        return MainActivityInjector.class;
    }
}

File: app/src/main/java/com/github/goldy1992/mp3player/client/activities/SplashScreenEntryActivity.kt
package com.github.goldy1992.mp3player.client.activities

import android.Manifest.permission
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AppCompatActivity
import com.github.goldy1992.mp3player.R
import com.github.goldy1992.mp3player.client.MediaBrowserAdapter
import com.github.goldy1992.mp3player.client.MediaBrowserConnectorCallback
import com.github.goldy1992.mp3player.client.PermissionGranted
import com.github.goldy1992.mp3player.client.PermissionsProcessor
import com.github.goldy1992.mp3player.commons.Constants
import com.github.goldy1992.mp3player.dagger.components.DaggerMediaActivityCompatComponent
import com.github.goldy1992.mp3player.service.MediaPlaybackServiceInjector
import org.apache.commons.lang3.exception.ExceptionUtils
import javax.inject.Inject

class SplashScreenEntryActivity : AppCompatActivity(), MediaBrowserConnectorCallback, PermissionGranted {
    private var mediaBrowserAdapter: MediaBrowserAdapter? = null
    private var permissionsProcessor: PermissionsProcessor? = null
    @get:VisibleForTesting
    @set:VisibleForTesting
    @Volatile
    var isSplashScreenFinishedDisplaying = false
    @get:VisibleForTesting
    @set:VisibleForTesting
    @Volatile
    var isPermissionGranted = false
    private var mainActivityIntent: Intent? = null
    @get:VisibleForTesting
    val thread = Thread(Runnable { init() })

    public override fun onCreate(savedInstanceState: Bundle?) {
        if (!isTaskRoot
                && intent.hasCategory(Intent.CATEGORY_LAUNCHER)
                && intent.action != null && intent.action == Intent.ACTION_MAIN) {
            finish()
            return
        }
        initialiseDependencies()
        super.onCreate(savedInstanceState)
        // TODO: have this injected so that a test implementation can be provided
        mainActivityIntent = Intent(applicationContext, mainActivityClass)
        setContentView(R.layout.splash_screen)
        thread.start()
    }

    override fun onStart() {
        super.onStart()
        Log.i(LOG_TAG, "onStart")
    }

    @Synchronized
    private fun splashScreenRun() { //    Log.i(LOG_TAG, "splashscreen run");
        try {
            wait(WAIT_TIME)
        } catch (ex: InterruptedException) {
            Log.e(LOG_TAG, ExceptionUtils.getMessage(ex.fillInStackTrace()))
            Thread.currentThread().interrupt()
        } finally {
            isSplashScreenFinishedDisplaying = true
            notifyAll()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) { //   Log.i(LOG_TAG, "permission result");
        var permissionIsGranted = false
        if (permissions.size > 0 && grantResults.size > 0) {
            for (i in permissions.indices) {
                if (permission.WRITE_EXTERNAL_STORAGE == permissions[i] &&
                        grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    permissionIsGranted = true
                }
            }
        }
        if (permissionIsGranted) {
            onPermissionGranted()
        } else {
            Toast.makeText(applicationContext, "Permission denied, please enable Storage permissions in settings in order to uer the app", Toast.LENGTH_LONG).show()
            finish()
        }
    }

    // MediaBrowserConnectorCallback
    override fun onConnected() {
        onProcessingComplete()
        Log.i(LOG_TAG, "hit on connected")
    }

    // MediaBrowserConnectorCallback
    override fun onConnectionSuspended() {}

    // MediaBrowserConnectorCallback
    override fun onConnectionFailed() {
        Log.i(LOG_TAG, "connection failed")
    }

    @Synchronized
    private fun onProcessingComplete() {
        Log.i(LOG_TAG, "processing complete")
        while (!isSplashScreenFinishedDisplaying || !isPermissionGranted) {
            try {
                wait(Constants.ONE_SECOND)
            } catch (ex: InterruptedException) {
                val error = ExceptionUtils.getMessage(ex.fillInStackTrace())
                Log.e(LOG_TAG, error)
                Thread.currentThread().interrupt()
            }
        }
        startMainActivity(mainActivityIntent)
    }

    private fun startMainActivity(mainActivityIntent: Intent?) {
        Log.i(LOG_TAG, "start main activity")
        startActivityForResult(mainActivityIntent, APP_TERMINATED)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == APP_TERMINATED) {
            if (mediaBrowserAdapter != null) {
                mediaBrowserAdapter!!.disconnect()
            }
            finish()
        }
    }

    override fun onPermissionGranted() {
        Log.i(LOG_TAG, "permission granted")
        isPermissionGranted = true
        createService()
        val r: Runnable = Thread(Runnable { mediaBrowserAdapter!!.connect() })
        runOnUiThread(r)
    }

    fun init() {
        Log.i(LOG_TAG, "reset")
        permissionsProcessor!!.requestPermission(permission.WRITE_EXTERNAL_STORAGE)
        val splashScreenWaitThread = Thread(Runnable { splashScreenRun() })
        splashScreenWaitThread.start()
    }

    @Inject
    fun setPermissionsProcessor(permissionsProcessor: PermissionsProcessor?) {
        this.permissionsProcessor = permissionsProcessor
    }

    @Inject
    fun setMediaBrowserAdapter(mediaBrowserAdapter: MediaBrowserAdapter?) {
        this.mediaBrowserAdapter = mediaBrowserAdapter
    }

    private fun createService() {
        startService(Intent(applicationContext, MediaPlaybackServiceInjector::class.java))
    }

    private fun initialiseDependencies() {
        val component = DaggerMediaActivityCompatComponent
                .factory()
                .create(applicationContext, "SPSH_SCRN_ACTVTY_WRKR", this)
        val splashScreenEntryActivityComponent = component.splashScreenEntryActivity().create(this, this)
        splashScreenEntryActivityComponent.inject(this)
    }

    val mainActivityClass: Class<*>
        get() = MainActivityInjector::class.java

    companion object {
        private const val LOG_TAG = "SPLSH_SCRN_ENTRY_ACTVTY"
        private const val WAIT_TIME = 3000L
        const val APP_TERMINATED = 0x78
    }
}